uaclient/entitlements/tests/test_fips.py

```diff
@@ -35,49 +38,78 @@
+    @mock.patch('os.getuid', return_value=0)
+    def test_can_enable_false_on_containers(self, m_getuid):
+        """On containers, can_enable is false and operational status n/a."""
+        op_status, op_status_details = self.entitlement.operational_status()
+        assert status.INAPPLICABLE == op_status
+        msg_cant_install = 'Cannot install FIPS on a container'
+        assert msg_cant_install == op_status_details
         with mock.patch('sys.stdout', new_callable=StringIO) as m_stdout:
-            self.assertTrue(entitlement.can_enable())
-        self.assertEqual('', m_stdout.getvalue())
+            assert False is self.entitlement.can_enable()
+        assert msg_cant_install + '\n' == m_stdout.getvalue()
 
 
-class TestFIPSEntitlementEnable(TestCase):
+class TestFIPSEntitlementEnable(BaseEnabledFIPSConfig):
 
-    @mock.patch('uaclient.util.get_platform_info')
-    @mock.patch('os.getuid', return_value=0)
-    def test_enable_configures_apt_sources_and_auth_files(
-            self, m_getuid, m_platform_info):
-        """When entitled, configure apt repo auth token, pinning and url."""
-        m_platform_info.return_value = 'xenial'
-        tmp_dir = self.tmp_dir()
-        cfg = config.UAConfig(cfg={'data_dir': tmp_dir})
-        cfg.write_cache('machine-token', FIPS_MACHINE_TOKEN)
-        cfg.write_cache('machine-access-fips', FIPS_RESOURCE_ENTITLED)
-        entitlement = FIPSEntitlement(cfg)
-        # Unset static affordance container check
-        entitlement.static_affordances = ()
-
-        with mock.patch('uaclient.apt.add_auth_apt_repo') as m_add_apt:
+    @mock.patch('uaclient.util.get_platform_info', return_value='xenial')
+    def test_enable_configures_apt_sources_and_auth_files_and_installs(
+            self, _platform_info):
+        """When entitled, configure apt repo, auth, pin and install debs."""
+
+        self.entitlement.can_enable = lambda: True
+
+        orig_exists = os.path.exists
+
+        def fake_exists(path):
+            if path in (apt.APT_METHOD_HTTPS_FILE, apt.APT_METHOD_HTTPS_FILE):
+                return False
+            else:
+                return orig_exists(path)
+
+        with mock.patch('uaclient.apt.add_auth_apt_repo') as m_add_apt_src:
             with mock.patch('uaclient.apt.add_ppa_pinning') as m_add_pinning:
-                self.assertTrue(entitlement.enable())
+                with mock.patch('uaclient.util.subp') as m_subp:
+                    with mock.patch('os.path.exists', side_effect=fake_exists):
+                        self.assertTrue(self.entitlement.enable())
 
-        add_apt_calls = [
+        add_apt_src_calls = [
             mock.call('/etc/apt/sources.list.d/ubuntu-fips-xenial.list',
                       'http://FIPS', 'TOKEN', None, 'APTKEY')]
         apt_pinning_calls = [
             mock.call('/etc/apt/preferences.d/ubuntu-fips-xenial',
                       'http://FIPS', 1001)]
+        install_packages = self.entitlement.packages
```

Comment:
This doesn't need to be a copy as much as it does in the actual code above, but if it isn't then we're less likely to catch unexpected mutations like those I mentioned above.

Reply:
+1 will resolve to separate this in a new branch that I will link here.

Reply:
Resolved these comments in https://github.com/CanonicalLtd/ubuntu-advantage-client/pull/263